x-pack/plugins/alerting_builtins/server/alert_types/geo_threshold/geo_threshold.ts

```diff
@@ -0,0 +1,270 @@
+      .value()
+  );
+}
+
+interface EntityMovementDescriptor {
+  entityName: string;
+  currLocation: {
+    location: number[];
+    shapeId: string;
+    date: string | null;
+    docId: string;
+  };
+  prevLocation: {
+    location: number[];
+    shapeId: string;
+    date: string | null;
+    docId: string;
+  };
+}
+
+export function getMovedEntities(
+  currLocationArr: LatestEntityLocation[],
+  prevLocationArr: LatestEntityLocation[],
+  trackingEvent: string
+): EntityMovementDescriptor[] {
+  return (
+    currLocationArr
+      // Check if shape has a previous location and has moved
+      .reduce(
+        (
+          accu: EntityMovementDescriptor[],
+          {
+            entityName,
+            shapeLocationId,
+            dateInShape,
+            location,
+            docId,
+          }: {
+            entityName: string;
+            shapeLocationId: string;
+            dateInShape: string | null;
+            location: number[];
+            docId: string;
+          }
+        ) => {
+          const prevLocationObj = prevLocationArr.find(
+            (locationObj: LatestEntityLocation) => locationObj.entityName === entityName
+          );
+          if (!prevLocationObj) {
+            return accu;
+          }
+          if (shapeLocationId !== prevLocationObj.shapeLocationId) {
+            accu.push({
+              entityName,
+              currLocation: {
+                location,
+                shapeId: shapeLocationId,
+                date: dateInShape,
+                docId,
+              },
+              prevLocation: {
+                location: prevLocationObj.location,
+                shapeId: prevLocationObj.shapeLocationId,
+                date: prevLocationObj.dateInShape,
+                docId: prevLocationObj.docId,
+              },
+            });
+          }
+          return accu;
+        },
+        []
+      )
+      // Do not track entries to or exits from 'other'
+      .filter((entityMovementDescriptor: EntityMovementDescriptor) =>
+        trackingEvent === 'entered'
+          ? entityMovementDescriptor.currLocation.shapeId !== OTHER_CATEGORY
+          : entityMovementDescriptor.prevLocation.shapeId !== OTHER_CATEGORY
+      )
+  );
+}
+
+export const getGeoThresholdExecutor = ({ logger: log }: { logger: Logger }) =>
+  async function ({
+    previousStartedAt: currIntervalStartTime,
+    startedAt: currIntervalEndTime,
+    services,
+    params,
+    alertId,
+    state,
+  }: {
+    previousStartedAt: Date | null;
+    startedAt: Date;
+    services: AlertServices;
+    params: {
+      index: string;
+      indexId: string;
+      geoField: string;
+      entity: string;
+      dateField: string;
+      trackingEvent: string;
+      boundaryType: string;
+      boundaryIndexTitle: string;
+      boundaryIndexId: string;
+      boundaryGeoField: string;
+      boundaryNameField?: string;
+    };
+    alertId: string;
+    state: AlertTypeState;
+  }) {
```

Comment:
you probably want to type the return value here to be `AlertTypeState`, since that what the function _should_ be returning; that way you'll get type checking on your return value.